modal/_container_io_manager.py

```diff
@@ -388,15 +392,18 @@ async def _heartbeat_handle_cancellations(self) -> bool:
             # response.cancel_input_event.terminate_containers is never set, the server gets the worker to handle it.
             input_ids_to_cancel = response.cancel_input_event.input_ids
             if input_ids_to_cancel:
+                # Avoid issuing cancellation signals for the same inputs repeatedly.
+                new_input_ids_to_cancel = [
+                    input_id for input_id in input_ids_to_cancel if input_id not in self._in_progress_canceled_inputs
+                ]
+                self._in_progress_canceled_inputs.extend(new_input_ids_to_cancel)
+
                 if self._max_concurrency > 1:
-                    current_input_ids_to_cancel = [
-                        input_id for input_id in input_ids_to_cancel if input_id in self.current_inputs
-                    ]
-                    await self._push_terminated_outputs(current_input_ids_to_cancel)
-                    for input_id in current_input_ids_to_cancel:
-                        self.current_inputs[input_id].cancel()
-
-                elif self.current_input_id and self.current_input_id in input_ids_to_cancel:
+                    for input_id in new_input_ids_to_cancel:
+                        if input_id in self.current_inputs:
```

Comment:
What if the container doesn't have the input yet? Does this swallow the cancellation forever

Reply:
Rehashed the logic to use the existence of the cancel callback which is deregistered when the cancellation is in progress.